plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -0,0 +1,293 @@
+const APPLICATION_JSON: string = 'application/json';
+
+const HEADER_KUBERNETES_CLUSTERS: string = 'X-Kubernetes-Clusters';
+
+/**
+ *
+ * @alpha
+ */
+const ERROR_INTERNAL_SERVER: number = 500;
+
+const CLUSTER_USER_NAME: string = 'backstage';
+
+/**
+ *
+ * @alpha
+ */
+export interface KubernetesProxyResponse {
+  code: number;
+  data: any;
+  cluster?: string;
+}
+
+/**
+ *
+ * @alpha
+ */
+interface KubernetesProxyClusters {
+  [key: string]: string;
+}
+
+/**
+ *
+ * @alpha
+ */
+export class KubernetesProxy {
+  constructor(protected readonly logger: Logger) {}
+
+  public async handleProxyRequest(
+    clusterSupplier: KubernetesClustersSupplier,
+    req: Request,
+  ): Promise<KubernetesProxyResponse> {
+    const requestedClusters = this.getKubernetesRequestedClusters(req);
+
+    if (Object.keys(requestedClusters).length < 1) {
+      this.logger.error(`No clusters found`);
+      throw new NotFoundError('No clusters found!');
+    }
+
+    const clusterDetails = await this.getClusterDetails(
+      clusterSupplier,
+      requestedClusters,
+    );
+
+    if (clusterDetails.length < 1) {
+      this.logger.error(`No clusters found`);
+      throw new NotFoundError('No clusters found!');
+    }
+
+    const responses = await Promise.all(
+      clusterDetails.map(async clusterDetail => {
+        const response = await this.makeRequestToCluster(clusterDetail, req);
+        return response;
+      }),
+    );
+
+    const data: { [key: string]: any } = {};
+    const codes: number[] = [];
+
+    responses.forEach(kpr => {
+      if (kpr.cluster) {
+        data[kpr.cluster] = kpr.data;
+        codes.push(kpr.code);
+      }
+    });
+
+    const code = this.getBestResponseCode(codes);
+
+    const res: KubernetesProxyResponse = {
+      code,
+      data,
+    };
+
+    return res;
+  }
+
+  private getKubernetesRequestedClusters(
+    req: Request,
+  ): KubernetesProxyClusters {
+    const encodedClusters: string =
+      req.header(HEADER_KUBERNETES_CLUSTERS) ?? '';
+
+    if (!encodedClusters) {
+      return {};
+    }
+
+    try {
+      const decodedClusters = Buffer.from(encodedClusters, 'base64').toString();
+      const clusters: KubernetesProxyClusters = JSON.parse(decodedClusters);
+      return clusters;
+    } catch (e: any) {
+      this.logger.debug(
+        `error with encoded cluster header: ${stringifyError(e)}`,
+      );
+    }
+    return {};
+  }
+
+  private async getClusterDetails(
+    clusterSupplier: KubernetesClustersSupplier,
+    krc: KubernetesProxyClusters,
+  ): Promise<ClusterDetails[]> {
+    const clusters = await clusterSupplier.getClusters();
+
+    const clusterNames = Object.keys(krc);
+
+    const clusterDetails = clusters.filter(c => clusterNames.includes(c.name));
+
+    const clusterDetailsAuth = clusterDetails.map(c => {
+      const cAuth: ClusterDetails = Object.assign(c, {
+        serviceAccountToken: krc[c.name],
+      });
+      return cAuth;
+    });
```

Comment:
object.assign mutates the first argument ... did you mean something like this?

```ts
return clusterDetails.map(c => ({
  ...c,
  serviceAccountToken: krc[c.name],
});
```

Reply:
Didn't touch on this part much, if I were going for immutability I would use Immer library, can I?